app/controllers/bankdetails/BankAccountHasIbanController.scala

```diff
@@ -0,0 +1,91 @@
+import forms.{EmptyForm, Form2, InvalidForm, ValidForm}
+import javax.inject.{Inject, Singleton}
+import models.bankdetails.Account._
+import models.bankdetails.{BankAccount,  BankAccountHasIban, BankDetails}
+import services.StatusService
+import uk.gov.hmrc.play.audit.http.connector.AuditConnector
+import utils.{AuthAction, StatusConstants}
+
+import scala.concurrent.Future
+
+@Singleton
+class BankAccountHasIbanController @Inject()(
+                                           val dataCacheConnector: DataCacheConnector,
+                                           val authAction: AuthAction,
+                                           val auditConnector: AuditConnector,
+                                           val statusService: StatusService
+                                         ) extends BankDetailsController {
+
+  def get(index: Int, edit: Boolean = false) = authAction.async{
+      implicit request =>
+        for {
+          bankDetails <- getData[BankDetails](request.credId, index)
+          status <- statusService.getStatus(request.amlsRefNumber, request.accountTypeId, request.credId)
+        } yield bankDetails match {
+          case Some(x@BankDetails(_, _, Some(BankAccount(_, Some(data@BankAccountHasIban(_)), _)), _, _, _, _)) if x.canEdit(status) =>
+            Ok(views.html.bankdetails.bank_account_account_has_iban(Form2[BankAccountHasIban](data), edit, index))
+          case Some(x) if x.canEdit(status) =>
+            Ok(views.html.bankdetails.bank_account_account_has_iban(EmptyForm, edit, index))
+          case _ => NotFound(notFoundView)
+        }
+  }
+
+  def post(index: Int, edit: Boolean = false) = authAction.async {
+      implicit request => {
+
+        lazy val sendAudit = for {
+          details <- OptionT(getData[BankDetails](request.credId, index))
+          result <- OptionT.liftF(auditConnector.sendEvent(audit.AddBankAccountEvent(details)))
+        } yield result
+
+        Form2[BankAccountHasIban](request.body) match {
+          case f: InvalidForm =>
+            Future.successful(BadRequest(views.html.bankdetails.bank_account_account_has_iban(f, edit, index)))
+          case ValidForm(_, data) =>
+            updateDataStrict[BankDetails](request.credId, index) { bd =>
+              bd.copy(
+                bankAccount = Option(bd.bankAccount.getOrElse(BankAccount(None, None, None)).hasIban(data)),
+                status = Some(if (edit) {
+                  StatusConstants.Updated
+                } else {
+                  StatusConstants.Added
+                })
+              )
+            }.flatMap { _ =>
+                lazy val redirect = if(data.hasIban) {
+                  Redirect(routes.BankAccountIbanController.get(index))
+                } else {
+                  Redirect(routes.BankAccountNonUKController.get(index))
+                }
+                (sendAudit map { _ =>
+                  redirect
+                }) getOrElse redirect
```

Comment:
Possible refactor here to make this easier to follow? Perhaps split out the above into methods that do a single purpose?

Reply:
I've removed the audit here, as (I think) it should only be triggered on completion of all bank details